<div style="
text-align:center;
background-color:#556B2F;
color:white;
padding:20px;
border-radius:10px;
font-family:Arial, sans-serif;">

<h1>🛒 ShopEase Data Analytics Project</h1>
<h3>Data Cleaning, Exploratory Data Analysis & Business Insights</h3>

</div>

## Step 1: Import Libraries and load data set

In [14]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
from datetime import datetime

# Display and chart settings
pd.set_option("display.max_columns", None)
sns.set_style("whitegrid")

In [5]:
df = pd.read_csv("shopease_raw_orders.csv")
print(df.shape)
df.head()

(10020, 16)


,OrderID,CustomerID,OrderDate,CustomerAge,Gender,City,Category,Product,Quantity,UnitPrice,Discount,PaymentMethod,OrderStatus,DeliveryDate,Rating,TotalAmount
0,18129,C0507,2026-02-01 23:39:00,68.0,Female,Alexandria,Electronics,Laptop,4,18000,0.0,Wallet,Delivered,2026-02-04 23:39:00,5.0,72000.0
1,11666,C0207,2026-03-03 10:37:00,21.0,Female,Cairo,Sports,Yoga Mat,2,600,0.1,Wallet,Delivered,2026-03-07 10:37:00,1.0,1080.0
2,12859,C1112,2026-05-23 17:57:00,23.0,Male,Aswan,Fashion,Shoes,-1,1500,10%,Bank Transfer,Delivered,2026-05-27 17:57:00,4.0,1350.0
3,19787,C1322,2026-01-29 07:51:00,23.0,Male,Tanta,Beauty,Hair Dryer,4,1300,0.1,Wallet,Delivered,2026-01-30 07:51:00,4.0,4680.0
4,16473,C0314,2026-06-23 20:17:00,62.0,Female,Tanta,Electronics,Laptop,1,18000,0.1,Mastercard,Delivered,2026-06-27 20:17:00,5.0,16200.0


## Step 2.Introduction

### Background
E-commerce businesses generate large amounts of order data every day, covering what customers buy, how they pay, where they live and whether their orders are delivered. When this data is recorded by many systems and people, it usually contains errors such as inconsistent spellings, mixed date formats, missing entries and invalid values. Raw data like this cannot be trusted for decisions until it is cleaned and examined.

### Business Context
ShopEase is an online retail business selling products in five categories: Electronics, Fashion, Home, Sports and Beauty. Orders come from six cities (Cairo, Giza, Alexandria, Mansoura, Tanta and Aswan) and are paid for by cash, Visa, Mastercard, wallet or bank transfer. Each order ends up Delivered, Pending or Cancelled. Management needs to understand sales performance, customer behavior and order fulfilment.

The dataset used in this project is a **synthetic (practice) dataset** of 10,020 order records taken from Kaggle. The findings therefore describe this dataset and not a real company.

### Problem Statement
ShopEase's raw orders data contains quality issues that make direct analysis unreliable. Without cleaning and structured exploration, the business cannot accurately measure revenue, identify its strongest products and cities, understand cancellations and delivery performance, or judge the effect of discounts.

### Aim
To clean the ShopEase raw orders dataset and carry out an end-to-end exploratory data analysis using Python, in order to identify sales trends, customer and product patterns, and order-fulfilment issues, and to convert them into business insights and recommendations.

### Objectives
1. Understand the structure and content of the dataset.
2. Detect and correct data-quality problems: missing values, duplicates, inconsistent text, mixed formats, invalid values and outliers.
3. Create useful new features, such as order month, delivery time and discount percentage.
4. Measure key performance indicators (KPIs) such as total revenue, number of orders, average order value and cancellation rate.
5. Analyze sales by time, category, product, city and payment method.
6. Examine order status, delivery time and customer ratings.
7. Apply descriptive statistics and correlation to support the findings.
8. Present the results with suitable visualizations.
9. Give practical recommendations that are supported by the analysis.

### Business Questions
1. How much revenue does ShopEase generate, and how does it change month by month?
2. Which categories and products contribute most to revenue and orders?
3. What percentage of orders is delivered, pending or cancelled, and does it differ by city, category or payment method?
4. Which cities and payment methods are used most?
5. Do discounts change the quantity ordered, and how much revenue do they give up?
6. How long does delivery take, and does it vary by city?
7. Are customer ratings related to delivery time or discount?
8. What do customer age and gender tell us about buying behavior?

## Step 3.Column description & quick inspection

In [7]:
df.columns

Index(['OrderID', 'CustomerID', 'OrderDate', 'CustomerAge', 'Gender', 'City',
       'Category', 'Product', 'Quantity', 'UnitPrice', 'Discount',
       'PaymentMethod', 'OrderStatus', 'DeliveryDate', 'Rating',
       'TotalAmount'],
      dtype='object')

In [8]:
df.info()
print("\nOriginal data types:")
df.dtypes

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10020 entries, 0 to 10019
Data columns (total 16 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   OrderID        10020 non-null  int64  
 1   CustomerID     10005 non-null  object 
 2   OrderDate      10020 non-null  object 
 3   CustomerAge    9900 non-null   float64
 4   Gender         10020 non-null  object 
 5   City           10020 non-null  object 
 6   Category       10020 non-null  object 
 7   Product        10000 non-null  object 
 8   Quantity       10020 non-null  int64  
 9   UnitPrice      10020 non-null  int64  
 10  Discount       9920 non-null   object 
 11  PaymentMethod  9970 non-null   object 
 12  OrderStatus    10020 non-null  object 
 13  DeliveryDate   7790 non-null   object 
 14  Rating         7728 non-null   float64
 15  TotalAmount    10020 non-null  float64
dtypes: float64(3), int64(3), object(10)
memory usage: 1.2+ MB

Original data types:


OrderID            int64
CustomerID        object
OrderDate         object
CustomerAge      float64
Gender            object
City              object
Category          object
Product           object
Quantity           int64
UnitPrice          int64
Discount          object
PaymentMethod     object
OrderStatus       object
DeliveryDate      object
Rating           float64
TotalAmount      float64
dtype: object

In [9]:
df.describe()

,OrderID,CustomerAge,Quantity,UnitPrice,Rating,TotalAmount
count,10020.000000,9900.000000,10020.000000,10020.000000,7728.000000,10020.000000
mean,14999.295908,44.381212,1.947605,1956.412176,3.958851,3308.779741
std,2888.444735,15.703142,1.100476,5403.755496,1.023657,7328.702149
min,10001.000000,-3.000000,-3.000000,-500.000000,-1.000000,200.000000
25%,12496.750000,31.000000,1.000000,650.000000,3.000000,855.000000
50%,14998.500000,44.000000,2.000000,900.000000,4.000000,1600.000000
75%,17501.250000,58.000000,3.000000,1500.000000,5.000000,2853.250000
max,20000.000000,150.000000,4.000000,150000.000000,6.000000,72000.000000


In [10]:
df.isnull().sum()

OrderID             0
CustomerID         15
OrderDate           0
CustomerAge       120
Gender              0
City                0
Category            0
Product            20
Quantity            0
UnitPrice           0
Discount          100
PaymentMethod      50
OrderStatus         0
DeliveryDate     2230
Rating           2292
TotalAmount         0
dtype: int64

## Step 4. 🧹Data Preprocessing

## Cleaning stage (part 1)

### text cleaning

In [11]:
text_columns = ["Gender", "City", "Category", "Product", "PaymentMethod", "OrderStatus"]

for col in text_columns:
    df[col] = df[col].str.strip()

for col in ["City", "Category", "Product", "PaymentMethod"]:
    df[col] = df[col].str.title()

df["Gender"] = df["Gender"].str.lower().map({
    "male": "Male", "m": "Male",
    "female": "Female", "f": "Female"
})

df["OrderStatus"] = (
    df["OrderStatus"].str.lower()
    .replace({"canceled": "cancelled"})
    .str.title()
)

for col in ["Gender", "City", "Category", "PaymentMethod", "OrderStatus"]:
    print(col, "->", df[col].nunique(), "unique values")

print("Gender missing:", df["Gender"].isnull().sum())
print("Unique products:", df["Product"].nunique())

Gender -> 2 unique values
City -> 6 unique values
Category -> 5 unique values
PaymentMethod -> 5 unique values
OrderStatus -> 3 unique values
Gender missing: 0
Unique products: 25


### Discount

In [12]:
def convert_discount(value):
    if pd.isnull(value):
        return np.nan
    text = str(value).strip()
    if text.endswith("%"):
        return float(text.replace("%", "")) / 100
    return float(text)

df["Discount"] = df["Discount"].apply(convert_discount)

print(df["Discount"].value_counts(dropna=False).sort_index())

Discount
0.00    3465
0.05    1940
0.10    2515
0.15    1511
0.20     489
NaN      100
Name: count, dtype: int64


### dates

In [15]:
def convert_order_date(value):
    text = str(value).strip()
    possible_formats = ["%Y-%m-%d %H:%M:%S", "%Y-%m-%d", "%d-%m-%Y", "%d/%m/%Y", "%b %d %Y"]
    for date_format in possible_formats:
        try:
            return datetime.strptime(text, date_format)
        except ValueError:
            continue
    return pd.NaT

df["OrderDate"] = df["OrderDate"].apply(convert_order_date)
df["OrderDate"] = pd.to_datetime(df["OrderDate"])
df["DeliveryDate"] = pd.to_datetime(df["DeliveryDate"])

print("Missing OrderDate:", df["OrderDate"].isnull().sum())
print("Missing DeliveryDate:", df["DeliveryDate"].isnull().sum())
print("Earliest order:", df["OrderDate"].min())
print("Latest order:", df["OrderDate"].max())

Missing OrderDate: 25
Missing DeliveryDate: 2230
Earliest order: 2026-01-01 00:00:00
Latest order: 2026-06-29 23:40:00


### missing values and invalid ages and ratings

In [16]:
df["PaymentMethod"] = df["PaymentMethod"].fillna("Unknown")
df["Product"] = df["Product"].fillna("Unknown")

df.loc[(df["CustomerAge"] < 0) | (df["CustomerAge"] > 100), "CustomerAge"] = np.nan
df["CustomerAge"] = df["CustomerAge"].fillna(df["CustomerAge"].median())

df["Discount"] = df["Discount"].fillna(0)

df.loc[(df["Rating"] < 1) | (df["Rating"] > 5), "Rating"] = np.nan

df["Delivery_Date_Missing_Flag"] = (
    (df["OrderStatus"] == "Delivered") & (df["DeliveryDate"].isnull())
)

print(df.shape)
print("Mean rating:", round(df["Rating"].mean(), 2))
print("Age median:", df["CustomerAge"].median())
print("Flagged orders:", df["Delivery_Date_Missing_Flag"].sum())

(10020, 17)
Mean rating: 3.96
Age median: 44.0
Flagged orders: 35


### final check for this stage

In [17]:
print(df.isnull().sum())
df.dtypes

OrderID                          0
CustomerID                      15
OrderDate                       25
CustomerAge                      0
Gender                           0
City                             0
Category                         0
Product                          0
Quantity                         0
UnitPrice                        0
Discount                         0
PaymentMethod                    0
OrderStatus                      0
DeliveryDate                  2230
Rating                        2307
TotalAmount                      0
Delivery_Date_Missing_Flag       0
dtype: int64


OrderID                                int64
CustomerID                            object
OrderDate                     datetime64[ns]
CustomerAge                          float64
Gender                                object
City                                  object
Category                              object
Product                               object
Quantity                               int64
UnitPrice                              int64
Discount                             float64
PaymentMethod                         object
OrderStatus                           object
DeliveryDate                  datetime64[ns]
Rating                               float64
TotalAmount                          float64
Delivery_Date_Missing_Flag              bool
dtype: object

## Cleaning stage (part 2)-- Fixing Invalid Values 

### Each product has one fixed price 

In [18]:
known_products = df[df["Product"] != "Unknown"]
standard_price = known_products.groupby("Product")["UnitPrice"].median()
print(standard_price.sort_values())

Product
Shampoo            250.0
Lamp               300.0
Mouse              450.0
Face Cream         450.0
T-Shirt            500.0
Yoga Mat           600.0
Football           650.0
Bedsheet           700.0
Bag                750.0
Headphones         850.0
Jeans              900.0
Makeup Kit         900.0
Keyboard           900.0
Perfume           1100.0
Jacket            1200.0
Blender           1200.0
Hair Dryer        1300.0
Dumbbells         1400.0
Shoes             1500.0
Tennis Racket     1600.0
Coffee Maker      1800.0
Running Shoes     1800.0
Smart Watch       2200.0
Chair             2500.0
Laptop           18000.0
Name: UnitPrice, dtype: float64


every product has one fixed price. In all rows with a valid price, the price equals that standard price exactly. So when a price is wrong, we know what it should be.

### Fix invalid prices

In [20]:
df["StandardPrice"] = df["Product"].map(standard_price)

invalid_price = (df["UnitPrice"] <= 0) | (df["UnitPrice"] > df["StandardPrice"] * 3)
print("Invalid prices:", invalid_price.sum())

df.loc[invalid_price, "UnitPrice"] = df.loc[invalid_price, "StandardPrice"]

print("Lowest price:", df["UnitPrice"].min())
print("Highest price:", df["UnitPrice"].max())

Invalid prices: 40
Lowest price: 250
Highest price: 18000


### Fix invalid quantities

Quantity = TotalAmount ÷ (UnitPrice × (1 − Discount))

In [21]:
invalid_quantity = df["Quantity"] <= 0
print("Invalid quantities:", invalid_quantity.sum())

calculated_quantity = df["TotalAmount"] / (df["UnitPrice"] * (1 - df["Discount"]))
rounded_quantity = calculated_quantity.round()

is_whole_number = (calculated_quantity - rounded_quantity).abs() < 0.01
is_realistic = rounded_quantity.between(1, 4)
can_recover = invalid_quantity & is_whole_number & is_realistic
print("Can be recovered:", can_recover.sum())

df.loc[can_recover, "Quantity"] = rounded_quantity[can_recover]
df.loc[invalid_quantity & ~can_recover, "Quantity"] = np.nan

print("Quantity missing now:", df["Quantity"].isnull().sum())
print(df["Quantity"].value_counts(dropna=False).sort_index())

Invalid quantities: 80
Can be recovered: 79
Quantity missing now: 1
Quantity
1.0    4355
2.0    2896
3.0    1419
4.0    1349
NaN       1
Name: count, dtype: int64


In [23]:
df_raw = pd.read_csv("shopease_raw_orders.csv")  

print(df_raw.shape)
print(df.shape[0] == df_raw.shape[0])

(10020, 16)
True


In [24]:
discount_was_missing = df_raw["Discount"].isnull()
print("Discount was missing in:", discount_was_missing.sum(), "rows")

estimated_discount = 1 - df["TotalAmount"] / (df["Quantity"] * df["UnitPrice"])
estimated_discount = estimated_discount.round(2)

valid_levels = [0, 0.05, 0.1, 0.15, 0.2]
can_estimate = discount_was_missing & estimated_discount.isin(valid_levels)
print("Can be estimated:", can_estimate.sum())

df.loc[can_estimate, "Discount"] = estimated_discount[can_estimate]

print(df["Discount"].value_counts().sort_index())

Discount was missing in: 100 rows
Can be estimated: 100
Discount
0.00    3499
0.05    1958
0.10    2539
0.15    1532
0.20     492
Name: count, dtype: int64


### Check that the numbers now agree

In [25]:
calculated_total = df["Quantity"] * df["UnitPrice"] * (1 - df["Discount"])
matches = np.isclose(calculated_total, df["TotalAmount"])

print("Rows checked:", calculated_total.notnull().sum())
print("Rows that match:", matches.sum())
print("Match percentage:", round(matches.sum() / calculated_total.notnull().sum() * 100, 1))

Rows checked: 10019
Rows that match: 9920
Match percentage: 99.0


The key idea
1. TotalAmount = Quantity × UnitPrice × (1 − Discount), and TotalAmount was fully valid.
2. Each product has a fixed standard price. 40 invalid prices (25 zero or negative, 15 extreme) were replaced with it.
3. 80 invalid quantities: 79 recovered from the bill, 1 set to missing.
4. 100 missing discounts estimated from the bill instead of assuming 0. Include the finding that assuming 0 would have been wrong for 66 of them.
5. The 99% consistency check, and that about 99 rows keep their recorded TotalAmount.

## Cleaning stage (part 3) Dates, Duplicates and Final Row Cleanup

### Delivery date earlier than order date

In [26]:
bad_delivery = df["DeliveryDate"] < df["OrderDate"]
print("Delivered before ordered:", bad_delivery.sum())
print(df.loc[bad_delivery, "OrderStatus"].value_counts())

days_gap = (df["DeliveryDate"] - df["OrderDate"]).dt.days
print(days_gap[~bad_delivery].describe().round(2))

Delivered before ordered: 25
OrderStatus
Delivered    25
Name: count, dtype: int64
count    7746.00
mean        3.55
std         1.52
min         1.00
25%         2.00
50%         3.00
75%         5.00
max         7.00
dtype: float64


In [27]:
df.loc[bad_delivery, "DeliveryDate"] = pd.NaT

all 25 bad rows are Delivered orders. Normal delivery takes 1 to 7 days, so these rows are errors. We can't tell whether the order date or the delivery date is wrong, so we don't guess. We set the delivery date to missing:

### Orders with no usable order date

In [28]:
missing_order_date = df["OrderDate"].isnull()
print("Missing OrderDate:", missing_order_date.sum())
print(df.loc[missing_order_date, "OrderStatus"].value_counts())

Missing OrderDate: 25
OrderStatus
Delivered    19
Pending       4
Cancelled     2
Name: count, dtype: int64


keep these rows. They still have a valid product, city and TotalAmount. They just can't be placed on a timeline. Charts grouped by month leave them out automatically. We don't invent dates. In this report, note that 25 orders (0.25%) are excluded from time-based analysis only.

### Exact duplicates

In [29]:
print("Exact duplicate rows:", df.duplicated().sum())
print("Repeated OrderIDs:", df["OrderID"].duplicated().sum())

df = df.drop_duplicates()

print("Rows now:", len(df))
print("Repeated OrderIDs left:", df["OrderID"].duplicated().sum())

Exact duplicate rows: 11
Repeated OrderIDs: 20
Rows now: 10009
Repeated OrderIDs left: 9


in the raw file there were 8 exact duplicates. cleaning made more visible, because rows that looked different (Cairo vs cairo) are now identical. This is why we remove duplicates after cleaning. Nine OrderIDs still repeat, so they must differ slightly.

### Look at the remaining repeated OrderIDs

In [30]:
repeated = df[df["OrderID"].duplicated(keep=False)].sort_values("OrderID")
repeated[["OrderID", "OrderDate", "PaymentMethod", "Rating", "TotalAmount"]]

,OrderID,OrderDate,PaymentMethod,Rating,TotalAmount
5675,10193,2026-04-28 12:32:00,Wallet,4.0,4400.0
3479,10193,2026-04-28 12:32:00,Wallet,NaN,4400.0
5580,10213,2026-06-12 19:33:00,Visa,NaN,285.0
5911,10213,2026-06-12 00:00:00,Visa,NaN,285.0
9102,10938,2026-02-14 00:00:00,Wallet,NaN,54000.0
4850,10938,2026-02-14 03:04:00,Wallet,NaN,54000.0
9351,12294,2026-02-24 04:03:00,Mastercard,5.0,2040.0
6925,12294,2026-02-24 04:03:00,Unknown,5.0,2040.0
656,14692,2026-04-07 00:00:00,Mastercard,4.0,1530.0
9604,14692,2026-04-07 18:16:00,Mastercard,4.0,1530.0


each pair is the same order recorded twice, one copy a little less complete. An OrderID identifies one order, so we keep one row per OrderID, preferably the more complete one.

### Keep the most complete copy

In [32]:
df["Info_Score"] = df.notnull().sum(axis=1)
df["Info_Score"] = df["Info_Score"] - (df["PaymentMethod"] == "Unknown") - (df["Product"] == "Unknown")

df = df.sort_values("Info_Score", ascending=False, kind="stable")
df = df.drop_duplicates(subset="OrderID", keep="first")
df = df.sort_index().reset_index(drop=True)

df = df.drop(columns=["Info_Score", "StandardPrice"])

df["Delivery_Date_Missing_Flag"] = (df["OrderStatus"] == "Delivered") & (df["DeliveryDate"].isnull())

print(df.shape)
print("Unique OrderIDs:", df["OrderID"].nunique())
print("Delivered orders without delivery date:", df["Delivery_Date_Missing_Flag"].sum())

(10000, 17)
Unique OrderIDs: 10000
Delivered orders without delivery date: 60


### Final check for this stage

In [33]:
missing = df.isnull().sum()
print(missing[missing > 0])
print(df["OrderStatus"].value_counts())

CustomerID        15
OrderDate         25
Quantity           1
DeliveryDate    2250
Rating          2301
dtype: int64
OrderStatus
Delivered    7810
Pending      1211
Cancelled     979
Name: count, dtype: int64


Column--	Missing--	Reason
1. CustomerID	--15 --	Real missing data
2. OrderDate --	25 --	Not available or impossible dates
3. Quantity	-- 1 -- Could not be recovered
4. DeliveryDate --	2,250-- 1,211 Pending + 979 Cancelled + 60 Delivered with no usable date
5. Rating--	2,301-- Mostly undelivered orders, plus invalid and unrated ones

1. Delivery dates: 25 delivered orders had a delivery date before the order date. All were 1 to 4 days off, and we couldn't tell which date was wrong, so the delivery date was set to missing.
2. Missing order dates: 25 orders kept, excluded from time-based analysis only.
3. Duplicates: 11 exact duplicates plus 9 repeated OrderIDs (the same order with small differences). We kept the most complete copy of each.
4. Row count: 10,020 before, 10,000 after (20 removed, 0.2%).